x/datadeal/keeper/certificate.go

```diff
@@ -0,0 +1,132 @@
+package keeper
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/medibloc/panacea-core/v2/types/assets"
+	"github.com/medibloc/panacea-core/v2/x/datadeal/types"
+)
+
+func (k Keeper) SubmitConsent(ctx sdk.Context, cert *types.Certificate) error {
+	unsignedCert := cert.UnsignedCertificate
+	if err := k.oracleKeeper.VerifySignature(ctx, unsignedCert, cert.Signature); err != nil {
+		return sdkerrors.Wrapf(types.ErrSubmitConsent, err.Error())
+	}
+
+	if err := k.oracleKeeper.VerifyOracle(ctx, unsignedCert.OracleAddress); err != nil {
+		return sdkerrors.Wrapf(types.ErrSubmitConsent, err.Error())
+	}
+
+	deal, err := k.GetDeal(ctx, unsignedCert.DealId)
+	if err != nil {
+		return sdkerrors.Wrapf(types.ErrSubmitConsent, "failed to get deal. %v", err)
+	} else if deal.Status != types.DEAL_STATUS_ACTIVE {
+		return sdkerrors.Wrapf(types.ErrSubmitConsent, "deal status is not ACTIVE")
+	}
+
+	if err := k.verifyExistCertificate(ctx, unsignedCert.DealId, unsignedCert.DataHash); err != nil {
+		return sdkerrors.Wrapf(types.ErrSubmitConsent, err.Error())
+	}
+
+	if err := k.SetCertificate(ctx, cert); err != nil {
+		return sdkerrors.Wrapf(types.ErrSubmitConsent, err.Error())
+	}
+
+	if err := k.sendReward(ctx, deal, unsignedCert); err != nil {
+		return sdkerrors.Wrapf(types.ErrSubmitConsent, err.Error())
+	}
+
+	if err := k.postProcessingOfDeal(ctx, deal); err != nil {
+		return sdkerrors.Wrapf(types.ErrSubmitConsent, err.Error())
+	}
+
+	return nil
+}
+
+func (k Keeper) verifyExistCertificate(ctx sdk.Context, dealID uint64, dataHash string) error {
+	existUnsignedCert, err := k.GetCertificate(ctx, dealID, dataHash)
+	if err != types.ErrCertificateNotFound {
+		if existUnsignedCert != nil {
+			return fmt.Errorf("already exist certificate. dataHash: %s", dataHash)
+		} else {
+			return err
+		}
+	}
+	return nil
+}
```

Comment:
I think we can simply return if the store has the certificate or not.
```go
func (k Keeper) isProvidedCert(ctx sdk.Context, dealID uint64, dataHash string) bool {
    store := ctx.KVStore(k.storeKey)
    return store.Has(types.GetCertificateKey(dealID, dataHash))
}
```

Reply:
It seems good. I will fix soon.

Reply:
fixed: https://github.com/medibloc/panacea-core/pull/533/commits/fec02e4707f772c7ada78e92414e0aa8d2ea5a30